pkg/server/plugin/datastore/sql/migration.go

```diff
@@ -27,6 +40,23 @@ func migrateDB(db *gorm.DB, dbType string, log hclog.Logger) (err error) {
 		return initDB(db, dbType, log)
 	}
 
+	// TODO related epic https://github.com/spiffe/spire/issues/1083
+	// We do not allow breaking changes in 0.X beyond version 0.9.0, but
+	// 1.0+ may have such changes. Thus iff Major version is 0, we are
+	// compatible. This logic will change in 1.0+.
+	// With 1.0 code, we will need a proper version check between code and db versions.
+	if codeVersion.Major != 0 {
+		err = sqlError.New("current migration code not compatible with %s release", codeVersion.String())
+		log.Error(err.Error())
```

Comment:
for your second comment, this is related to #1138; ideally as the error propagates up the stack it will be wrapped with details at each level, but each level of that stack doesn't have structured field details of the other levels, and so only they can provide such context and so they should log as well. Additionally a called module should not depend on knowing whether the caller will log its errors.